libs/chain-signatures/contract/src/lib.rs

```diff
@@ -750,23 +709,32 @@ impl VersionedMpcContract {
     /// 3. `return_signature_and_clean_state_on_success` fails with `ContractSignatureRequest`
     ///    argument.
     #[private]
-    #[handle_result]
     pub fn return_signature_and_clean_state_on_success(
         &mut self,
         request: SignatureRequest, // this change here should actually be ok.
         #[callback_result] signature: Result<SignatureResponse, PromiseError>,
-    ) -> Result<SignatureResponse, Error> {
+    ) -> PromiseOrValue<SignatureResponse> {
         let Self::V0(mpc_contract) = self;
+        mpc_contract.pending_requests.remove(&request);
         match signature {
-            Ok(signature) => {
-                log!("Signature is ready.");
-                mpc_contract.pending_requests.remove(&request);
-                Ok(signature)
+            Ok(signature) => PromiseOrValue::Value(signature),
+            Err(_) => {
+                PromiseOrValue::Promise(Promise::new(env::current_account_id()).function_call(
+                    "fail_on_timeout".to_string(),
+                    vec![],
+                    NearToken::from_near(0),
+                    Gas::from_tgas(1),
+                ))
             }
-            Err(_) => Err(SignError::Timeout.into()),
         }
     }
 
+    #[private]
+    pub fn fail_on_timeout(&self) {
+        // To stay consistent with the old version of the timeout error
+        panic!("{}", SignError::Timeout.to_string());
```

Comment:
What's the diff between `env::panic_str` and `panic!`?

Reply:
panic_str is slightly better because it doesn't bring in `String` dependencies, but we already use `format!` in other places, so realistically it is not different